DInstanceDs with explicit foralls no longer sweeten correctly under GHC HEAD (9.1)

I discovered that the `th-desugar` test suite fails to compile on recent versions of GHC HEAD. Here is a somewhat minimized example of the phenomenon:

```hs
{-# LANGUAGE TemplateHaskell #-}
{-# OPTIONS_GHC -ddump-splices #-}
module Bug where

import Language.Haskell.TH
import Language.Haskell.TH.Desugar

class C a
data T a

$(do let a = mkName "a"
         dec = decToTH $ DInstanceD Nothing (Just [DPlainTV a ()]) [] (DConT ''C `DAppT` (DConT ''T `DAppT` DVarT a)) []
     pure [dec])
```

This compiles on GHC 9.0.1-alpha1 and earlier, but not on HEAD:

```
$ ~/Software/ghc-9.1.0.20201217/bin/ghc Bug.hs
Loaded package environment from /home/ryanglscott/Documents/Hacking/Haskell/staging/.ghc.environment.x86_64-linux-9.1.0.20201217
[1 of 1] Compiling Bug              ( Bug.hs, Bug.o, Bug.dyn_o )
Bug.hs:(11,2)-(13,16): Splicing declarations
    do let a_a12Y = mkName "a"
           dec_a12Z
             = decToTH
                 $ DInstanceD
                     Nothing (Just [DPlainTV a_a12Y ()]) []
                     (DConT ''C `DAppT` (DConT ''T `DAppT` DVarT a_a12Y)) []
       pure [dec_a12Z]
  ======>
    instance forall a. C (T a)

Bug.hs:11:2: error:
    Instance head cannot contain nested ‘forall’s or contexts
    In an instance declaration
   |
11 | $(do let a = mkName "a"
   |  ^^^^^^^^^^^^^^^^^^^^^^...
```

Ultimately, I only have myself to blame here, since I simultaneously [added this feature to `th-desugar`](https://github.com/goldfirere/th-desugar/issues/118) and later [made it illegal in GHC with this patch](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4107) that restricts the use of nested `forall`s in instance heads. Apparently, the use of something like `` InstanceD Nothing [] (ForallT [PlainTV a ()] [] (ConT ''C `AppT` (ConT ''T `AppT` VarT a)) [] `` is close enough to a nested `forall` that GHC rejects it (even though it pretty-prints it as `instance forall a. C (T a)`). Sigh.

I now have to figure out how to dig my way out of the mess I created. Some possibilities:

* Patch GHC upstream so that `` InstanceD Nothing [] (ForallT [PlainTV a ()] [] (ConT ''C `AppT` (ConT ''T `AppT` VarT a)) [] `` doesn't trigger a nested `forall` warning.
* Remove the `Maybe [DTyVarBndr]` field from `DInstanceD`.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -15,6 +15,12 @@ Version 1.13 [????.??.??]
   +  | DConP Name [DType] [DPat] -- fun (Just @t x) = ...
      | ...
   ```
+* The `Maybe [DTyVarBndrUnit]` fields in `DInstanceD` and `DStandaloneDerivD`
+  are no longer used when sweetening. Previously, `th-desugar` would attempt to
+  sweeten these `DTyVarBndrUnit`s by turning them into a nested `ForallT`, but
+  GHC 9.2 or later no longer allow this, as they forbid nested `forall`s in
+  instance heads entirely. As a result, the `Maybe [DTyVarBndrUnit]` fields are
+  now only useful for functions that consume `DDec`s directly.
 
 Version 1.12 [2021.03.12]
 -------------------------

--- a/Language/Haskell/TH/Desugar/AST.hs
+++ b/Language/Haskell/TH/Desugar/AST.hs
@@ -113,6 +113,9 @@ data DDec = DLetDec DLetDec
           | DDataD NewOrData DCxt Name [DTyVarBndrUnit] (Maybe DKind) [DCon] [DDerivClause]
           | DTySynD Name [DTyVarBndrUnit] DType
           | DClassD DCxt Name [DTyVarBndrUnit] [FunDep] [DDec]
+            -- | Note that the @Maybe [DTyVarBndrUnit]@ field is dropped
+            -- entirely when sweetened, so it is only useful for functions
+            -- that directly consume @DDec@s.
           | DInstanceD (Maybe Overlap) (Maybe [DTyVarBndrUnit]) DCxt DType [DDec]
           | DForeignD DForeign
           | DOpenTypeFamilyD DTypeFamilyHead
@@ -122,6 +125,9 @@ data DDec = DLetDec DLetDec
                        [DCon] [DDerivClause]
           | DTySynInstD DTySynEqn
           | DRoleAnnotD Name [Role]
+            -- | Note that the @Maybe [DTyVarBndrUnit]@ field is dropped
+            -- entirely when sweetened, so it is only useful for functions
+            -- that directly consume @DDec@s.
           | DStandaloneDerivD (Maybe DDerivStrategy) (Maybe [DTyVarBndrUnit]) DCxt DType
           | DDefaultSigD Name DType
           | DPatSynD Name PatSynArgs DPatSynDir DPat

--- a/Language/Haskell/TH/Desugar/Sweeten.hs
+++ b/Language/Haskell/TH/Desugar/Sweeten.hs
@@ -43,7 +43,7 @@ import Language.Haskell.TH hiding (cxt)
 import Language.Haskell.TH.Datatype.TyVarBndr
 
 import Language.Haskell.TH.Desugar.AST
-import Language.Haskell.TH.Desugar.Core (DTypeArg(..), changeDTVFlags)
+import Language.Haskell.TH.Desugar.Core (DTypeArg(..))
 import Language.Haskell.TH.Desugar.Util
 
 expToTH :: DExp -> Exp
@@ -112,19 +112,9 @@ decToTH (DDataD Newtype _cxt _n _tvbs _mk _cons _derivings) =
 decToTH (DTySynD n tvbs ty) = TySynD n (map tvbToTH tvbs) (typeToTH ty)
 decToTH (DClassD cxt n tvbs fds decs) =
   ClassD (cxtToTH cxt) n (map tvbToTH tvbs) fds (decsToTH decs)
-decToTH (DInstanceD over mtvbs _cxt _ty decs) =
-  instanceDToTH over cxt' ty' decs
-  where
-  (cxt', ty') = case fmap (changeDTVFlags SpecifiedSpec) mtvbs of
-                  Nothing    -> (_cxt, _ty)
-                  Just _tvbs ->
-#if __GLASGOW_HASKELL__ < 800 || __GLASGOW_HASKELL__ >= 802
-                                ([], DForallT (DForallInvis _tvbs) $ DConstrainedT _cxt _ty)
-#else
-                                -- See #117
-                                error $ "Explicit foralls in instance declarations "
-                                     ++ "are broken on GHC 8.0."
-#endif
+decToTH (DInstanceD over _mtvbs cxt ty decs) =
+  -- We deliberately avoid sweetening _mtvbs. See #151.
+  instanceDToTH over cxt ty decs
 decToTH (DForeignD f) = ForeignD (foreignToTH f)
 #if __GLASGOW_HASKELL__ > 710
 decToTH (DOpenTypeFamilyD (DTypeFamilyHead n tvbs frs ann)) =
@@ -161,19 +151,9 @@ decToTH (DClosedTypeFamilyD (DTypeFamilyHead n tvbs frs _ann) eqns) =
   ClosedTypeFamilyD n (map tvbToTH tvbs) (frsToTH frs) (map (snd . tySynEqnToTH) eqns)
 #endif
 decToTH (DRoleAnnotD n roles) = RoleAnnotD n roles
-decToTH (DStandaloneDerivD mds mtvbs _cxt _ty) =
-  standaloneDerivDToTH mds cxt' ty'
-  where
-  (cxt', ty') = case fmap (changeDTVFlags SpecifiedSpec) mtvbs of
-                  Nothing    -> (_cxt, _ty)
-                  Just _tvbs ->
-#if __GLASGOW_HASKELL__ < 710 || __GLASGOW_HASKELL__ >= 802
-                                ([], DForallT (DForallInvis _tvbs) $ DConstrainedT _cxt _ty)
-#else
-                                -- See #117
-                                error $ "Explicit foralls in standalone deriving declarations "
-                                     ++ "are broken on GHC 7.10 and 8.0."
-#endif
+decToTH (DStandaloneDerivD mds _mtvbs cxt ty) =
+  -- We deliberately avoid sweetening _mtvbs. See #151.
+  standaloneDerivDToTH mds cxt ty
 #if __GLASGOW_HASKELL__ < 709
 decToTH (DDefaultSigD {})      =
   error "Default method signatures supported only in GHC 7.10+"

--- a/Test/Dec.hs
+++ b/Test/Dec.hs
@@ -45,16 +45,12 @@ $(S.dectest14)
 $(S.dectest15)
 #endif
 
-{-
-Temporarily disabled until #151 is addressed
-
 #if __GLASGOW_HASKELL__ < 800 || __GLASGOW_HASKELL__ >= 802
 $(S.dectest16)
 #endif
 #if __GLASGOW_HASKELL__ >= 802
 $(S.dectest17)
 #endif
--}
 
 #if __GLASGOW_HASKELL__ >= 809
 $(S.dectest18)

--- a/Test/DsDec.hs
+++ b/Test/DsDec.hs
@@ -69,16 +69,12 @@ $(dsDecSplice S.dectest14)
 $(dsDecSplice S.dectest15)
 #endif
 
-{-
-Temporarily disabled until #151 is addressed
-
 #if __GLASGOW_HASKELL__ < 800 || __GLASGOW_HASKELL__ >= 802
 $(return $ decsToTH [S.ds_dectest16])
 #endif
 #if __GLASGOW_HASKELL__ >= 802
 $(return $ decsToTH [S.ds_dectest17])
 #endif
--}
 
 #if __GLASGOW_HASKELL__ >= 809
 $(dsDecSplice S.dectest18)

--- a/Test/Splices.hs
+++ b/Test/Splices.hs
@@ -453,6 +453,14 @@ class ExCls a
 data ExData1 a
 data ExData2 a
 
+-- ds_dectest{16,17} demonstrate instance declarations with outermost foralls,
+-- a feature which Template Haskell itself does not yet support (see #151).
+-- For this reason, the closest we can get to this in TH is to construct
+-- equivalent Decs, dectest{16,17}, that drop the outermost foralls. The test
+-- suite ensures that this process happens automatically during sweetening by
+-- checking that the sweetened versions of ds_dectest{16,17} equal
+-- dectest{16,17}.
+
 ds_dectest16 = DInstanceD Nothing (Just [DPlainTV (mkName "a") ()]) []
                 (DConT ''ExCls `DAppT`
                   (DConT ''ExData1 `DAppT` DVarT (mkName "a"))) []
@@ -461,9 +469,8 @@ dectest16 = return [ InstanceD
 #if __GLASGOW_HASKELL__ >= 800
                        Nothing
 #endif
-                       [] (ForallT [plainTVSpecified (mkName "a")] []
-                                   (ConT ''ExCls `AppT`
-                                     (ConT ''ExData1 `AppT` VarT (mkName "a")))) [] ]
+                       [] (ConT ''ExCls `AppT`
+                            (ConT ''ExData1 `AppT` VarT (mkName "a"))) [] ]
 ds_dectest17 = DStandaloneDerivD Nothing (Just [DPlainTV (mkName "a") ()]) []
                 (DConT ''ExCls `DAppT`
                   (DConT ''ExData2 `DAppT` DVarT (mkName "a")))
@@ -473,9 +480,8 @@ dectest17 = return [ StandaloneDerivD
 #if __GLASGOW_HASKELL__ >= 802
                        Nothing
 #endif
-                       [] (ForallT [plainTVSpecified (mkName "a")] []
-                                   (ConT ''ExCls `AppT`
-                                     (ConT ''ExData2 `AppT` VarT (mkName "a")))) ]
+                       [] (ConT ''ExCls `AppT`
+                            (ConT ''ExData2 `AppT` VarT (mkName "a"))) ]
 #endif
 
 #if __GLASGOW_HASKELL__ >= 809

